### PR TITLE
render: Implement PixelBender Operation::Select (renamed from Loop)

### PR DESCRIPTION
**Pull request by Aaron1011**
This is not actually an loop opcode - it's a 'ternary if' opcode, which selects between two input registers based on a condition register, and writes to a destination register.

Fixes https://github.com/ruffle-rs/ruffle/issues/14498

### CODE REVIEW

**Comment by evilpie** [comment]
I would just call it  `select` like llvm or mesa.

**Comment by torokati44** [comment]
It's strange that `pbj` thinks it's a loop opcode[1], and `pbjas` simply skips it[2].

I guess it's not used that often...?

 [1]: https://github.com/HaxeFoundation/format/blob/39787764801f9e02c5b5ed771490e767a5488e65/format/pbj/Reader.hx#L258-L259
 [2]: https://github.com/jamesward/pbjas/blob/871e66d6289a089df6dc7511ef8921100a5bd630/pbjAS/pbjAS/PBJAssembler.as#L527-L546

File: tests/tests/swfs/avm2/pixelbender_dithering/test.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Based on  https://github.com/FlashBacks1998/Custom-Pixel-Bender-Example-Ordered-Dithering-
```
Comment (torokati44) [inline tests/tests/swfs/avm2/pixelbender_dithering/test.toml:1]: absolute nit: double space

File: tests/tests/swfs/avm2/pixelbender_dithering/test.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+# The prints were sorted to avoid relying on object iteration order
+num_ticks = 1
+
+[image_comparisons.output]
+tolerance = 1
+max_outliers = 19813
```
Comment (torokati44) [inline tests/tests/swfs/avm2/pixelbender_dithering/test.toml:7]: That's... um... a lot?
Perhaps a bit higher tolerance could be considered instead?

**Comment by torokati44** [comment]
@evilpie
> I would just call it select like llvm or mesa.

I'm reading in at least two places (https://github.com/eclipse/omr/pull/668#issuecomment-273605969 and https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Control_flow/Select) that in general, the difference between "ternary if" and "select" is that the former only evaluates the single branch whose value is to be returned, while the latter evaluates both (or all, in the general case) branches, and _selects_ a value to return afterwards.

Perhaps we should follow this convention with the naming - that is, if the difference is observable in PB at all.

**Comment by lvyitian** [comment]
In https://web.archive.org/web/20100223210221/http://www.kaourantin.net/source/pbjtools/apbj.cpp, opcode 0x33 is called "select".

**Comment by torokati44** [comment]
Ooh, nice find! The entire blog archive in fact!

**Comment by Aaron1011** [comment]
I had based my changes off of `apbj.cpp`, but I preferred `TernaryIf` to `Select`. I can change it to `Select` to follow the convention, though (the opcode uses PixelBender registers, so both 'arguments' are always evaluated)

**Comment by torokati44** [comment]
I'd say that's two reasons for going with `Select` then.

**Comment by torokati44** [comment]
We were talking about `Select`, not `Switch`. 😶

**Comment by Aaron1011** [comment]
Oops - I don't know why I did `Swich`

File: render/naga-pixelbender/src/lib.rs [hunk 3]
```diff
@@ -1388,8 +1388,84 @@ impl<'a> ShaderBuilder<'a> {
...
+                } => {
+                    let src1_expr = self.load_src_register(src1)?;
+                    let src2_expr = self.load_src_register(src2)?;
+
+                    let expr_zero: Handle<Expression> = match condition.kind {
+                        PixelBenderRegKind::Float => self.zerof32,
+                        PixelBenderRegKind::Int => self.zeroi32,
+                    };
+                    if condition.channels.len() != 1 {
+                        panic!("Ternary if condition must be a scalar: {condition:?}");
```
Comment (evilpie) [inline render/naga-pixelbender/src/lib.rs:1405]: s/ternary if/select/

**Comment by evilpie** [comment]
I was wondering if we could use the https://docs.rs/naga/latest/naga/enum.Expression.html#variant.Select?

**Comment by Aaron1011** [comment]
@evilpie Good idea - I've updated the pr

**Comment by evilpie** [comment]
LGTM

**Comment by torokati44** [comment]
I'd still like a bit higher tolerance and a lot fewer accepted outliers for the test.

**Comment by Aaron1011** [comment]
Even with a toleranece of 10, there are still 19654 outliers. I'd really like to get this merged, and work on improving the tests as a follow-up.

**Comment by torokati44** [comment]
That's fine by me.

**Review by Lord-McSweeney — APPROVED** [review]

**Comment by torokati44** [comment]
To anyone who will end up merging - please squash! :)

**Pull request merged**